### PR TITLE
cache JWTs

### PR DESCRIPTION
**Pull request by sahil-lakhwani**
- Caches JWTs
- Addresses https://github.com/infracloudio/msbotbuilder-go/issues/9

### CODE REVIEW

File: connector/client/client.go [hunk 1]
```diff
@@ -78,42 +93,51 @@ func (client ConnectorClient) Post(target url.URL, activity schema.Activity) err
 
 func (client *ConnectorClient) getToken() (string, error) {
 
-	data := url.Values{}
-	data.Set("grant_type", "client_credentials")
-	data.Set("client_id", client.Credentials.GetAppID())
-	data.Set("client_secret", client.Credentials.GetAppPassword())
-	data.Set("scope", auth.ToChannelFromBotOauthScope)
+	if cache == nil || cache.IsExpired() {
```
Comment (PrasadG193) [inline connector/client/client.go]: It's little difficult to understand. Let's change this condition `if !cache.IsExpired()` and return `cache.token` here

File: connector/client/client.go [hunk 2]
```diff
@@ -9,12 +9,27 @@ import (
...
 
 	"github.com/infracloudio/msbotbuilder-go/connector/auth"
 	"github.com/infracloudio/msbotbuilder-go/schema"
 	"github.com/infracloudio/msbotbuilder-go/schema/customerror"
 )
 
+type jwtCache struct {
+	token  string
+	Expiry time.Time
+}
```
Comment (PrasadG193) [inline connector/client/client.go]: @sahil-lakhwani let's put cache related changes in separate `cache.go` file

File: connector/client/client.go [hunk 3]
```diff
@@ -9,12 +9,15 @@ import (
...
 	"strconv"
 	"strings"
+	"time"
 
 	"github.com/infracloudio/msbotbuilder-go/connector/auth"
 	"github.com/infracloudio/msbotbuilder-go/schema"
 	"github.com/infracloudio/msbotbuilder-go/schema/customerror"
 )
 
+var cache *jwtCache
```
Comment (PrasadG193) [inline connector/client/client.go]: @sahil-lakhwani let's follow the same approach as of JWK cache and get rid of this singleton variable

File: connector/client/cache.go [hunk 4]
```diff
@@ -0,0 +1,16 @@
+package client
+
+import "time"
+
+type jwtCache struct {
+	token  string
+	Expiry time.Time
```
Comment (PrasadG193) [inline connector/client/cache.go]: `Export` needs to be exported?

File: core/bot_framework_adapter.go [hunk 5]
```diff
@@ -46,7 +47,18 @@ func NewBotAdapter(settings AdapterSetting) Adapter {
 	if settings.ChannelService == "" {
 		settings.ChannelService = auth.ChannelService
 	}
-	return &BotFrameworkAdapter{settings}
+
+	clientConfig, err := client.NewClientConfig(settings.CredentialProvider, auth.ToChannelFromBotLoginURL[0])
+	if err != nil {
+		return nil
```
Comment (PrasadG193) [inline core/bot_framework_adapter.go]: Please return error here. Let's change return type of `NewBotAdapter`

File: connector/client/cache.go [hunk 6]
```diff
@@ -0,0 +1,16 @@
+package client
+
+import "time"
+
+type jwtCache struct {
```
Comment (PrasadG193) [inline connector/client/cache.go]: Please write one liner description.

**Comment by PrasadG193** [comment]
@sahil-lakhwani Now, since we have decided to make JWT and JWK cache a part of TokenValidator and ConnectorClient. Can we move cache to a separate package?

**Comment by sahil-lakhwani** [comment]
Closing this in favor of https://github.com/infracloudio/msbotbuilder-go/pull/24

**Pull request closed without merging**